File: platform/polyglot/confluence_api/py/src/confluence_api/services/group_service.py
"""
Group service for Confluence Data Center REST API v9.2.3.

Provides operations for retrieving groups and group members.
"""

from __future__ import annotations

from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence_api", __file__)


class GroupService:
    """Service for Confluence group operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    def get_group(self, group_name: str) -> dict:
        """Retrieve a single group by name."""
        log.debug("get_group called", {"group_name": group_name})
        result = self._client.get(f"group/{group_name}")
        log.info("get_group succeeded", {"group_name": group_name})
        return result

    def get_groups(self, start: int = 0, limit: int = 25) -> dict:
        """Retrieve a paginated list of groups."""
        log.debug("get_groups called", {"start": start, "limit": limit})
        params: dict[str, Any] = {"start": start, "limit": limit}
        result = self._client.get("group", params=params)
        log.info("get_groups succeeded", {"start": start, "limit": limit})
        return result

    def get_group_members(
        self,
        group_name: str,
        start: int = 0,
        limit: int = 25,
    ) -> dict:
        """Retrieve a paginated list of members in a group."""
        log.debug("get_group_members called", {
            "group_name": group_name,
            "start": start,
            "limit": limit,
        })
        params: dict[str, Any] = {"start": start, "limit": limit}
        result = self._client.get(f"group/{group_name}/member", params=params)
        log.info("get_group_members succeeded", {"group_name": group_name})
        return result


class AsyncGroupService:
    """Async service for Confluence group operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def get_group(self, group_name: str) -> dict:
        """Retrieve a single group by name."""
        log.debug("get_group called", {"group_name": group_name})
        result = await self._client.get(f"group/{group_name}")
        log.info("get_group succeeded", {"group_name": group_name})
        return result

    async def get_groups(self, start: int = 0, limit: int = 25) -> dict:
        """Retrieve a paginated list of groups."""
        log.debug("get_groups called", {"start": start, "limit": limit})
        params: dict[str, Any] = {"start": start, "limit": limit}
        result = await self._client.get("group", params=params)
        log.info("get_groups succeeded", {"start": start, "limit": limit})
        return result

    async def get_group_members(
        self,
        group_name: str,
        start: int = 0,
        limit: int = 25,
    ) -> dict:
        """Retrieve a paginated list of members in a group."""
        log.debug("get_group_members called", {
            "group_name": group_name,
            "start": start,
            "limit": limit,
        })
        params: dict[str, Any] = {"start": start, "limit": limit}
        result = await self._client.get(f"group/{group_name}/member", params=params)
        log.info("get_group_members succeeded", {"group_name": group_name})
        return result

File: platform/polyglot/confluence_api/mjs/src/services/group-service.mjs
/**
 * @module services/group-service
 * @description Service for Confluence Group REST API operations.
 *
 * Provides read access to groups and group membership.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence group operations.
 */
export class GroupService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  /**
   * Get a single group by name.
   *
   * @param {string} groupName - The group name.
   * @returns {Promise<Object>} The group object.
   */
  async getGroup(groupName) {
    log.debug('getGroup called', { groupName });
    try {
      const result = await this._client.get(`group/${groupName}`);
      log.info('getGroup succeeded', { groupName, name: result.name });
      return result;
    } catch (error) {
      log.error('getGroup failed', { groupName, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get a paginated list of all groups.
   *
   * @param {Object} [options={}] - Query options.
   * @param {number} [options.start=0] - Pagination start offset.
   * @param {number} [options.limit=25] - Maximum number of results.
   * @returns {Promise<Object>} Paginated group list.
   */
  async getGroups({ start = 0, limit = 25 } = {}) {
    log.debug('getGroups called', { start, limit });
    try {
      const queryParams = { start, limit };

      const result = await this._client.get('group', { queryParams });
      log.info('getGroups succeeded', { size: result.size ?? result.results?.length });
      return result;
    } catch (error) {
      log.error('getGroups failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get members of a group.
   *
   * @param {string} groupName - The group name.
   * @param {Object} [options={}] - Query options.
   * @param {number} [options.start=0] - Pagination start offset.
   * @param {number} [options.limit=25] - Maximum number of results.
   * @returns {Promise<Object>} Paginated group members list.
   */
  async getGroupMembers(groupName, { start = 0, limit = 25 } = {}) {
    log.debug('getGroupMembers called', { groupName, start, limit });
    try {
      const queryParams = { start, limit };

      const result = await this._client.get(`group/${groupName}/member`, { queryParams });
      log.info('getGroupMembers succeeded', { groupName, size: result.size ?? result.results?.length });
      return result;
    } catch (error) {
      log.error('getGroupMembers failed', { groupName, message: error.message, status: error.status });
      throw error;
    }
  }
}
